feat: Implement Python version of `memorywipe.sh` script

p1utoze opened the pull request:
This is a single-file python script that implements most of the major commands used in the original bash script. 
I've used `click` library instead of the built-in `argparse` or other equivalent. Here's the reason behind it- [link](https://click.palletsprojects.com/en/8.1.x/why/)

closes #5 

**Changes made**:
- I've nested the high-level commands: `sanitization`, `extraction`, and `verify` under main() group
- Callback functions for validation of the above commands (except verify)
- The users can either execute the command without option that triggers an interactive mode
```
python3 memorywipe.sh sanitize
```
or pass the optional flag values directly.
- Used subprocess library to execute shell commands. 
- Defined class for sanitize command

p1utoze committed:
feat: add top-level commands in main.py

p1utoze committed:
Merge pull request #1 from p1utoze/py-codebase

feat: add top-level commands in main.py

p1utoze committed:
update: added sanitize menu callback and Sanitization class

p1utoze committed:
update: add extract command functionality and callbacks in memorywipe.py

p1utoze committed:
Merge branch 'SuchitReddi:main' into main

p1utoze committed:
Merge branch 'main' of https://github.com/p1utoze/memorywipe into py-codebase

p1utoze committed:
update: add cryptographic wipe functionality for sanitize command

p1utoze committed:
update: add placeholder functions temporarily in ata_hdparm method

p1utoze committed:
update: add partition validation function in memorywipe.py

p1utoze committed:
update: added required methods for ata_hdparm erase method

p1utoze committed:
update: implement auto wipe method to Sanitization class

p1utoze committed:
refactor: use in-built prompting parameters for input

SuchitReddi commented:
Good work! I will review it. Have you tested it on any device?

p1utoze commented:
I'll try using a USB drive tomorrow. While SSDs were the primary target for this tool, it's difficult to get one at the moment for me.

SuchitReddi commented:
It is not a requirement to test only on SSDs. You can proceed with testing on USB.

p1utoze committed:
update: add bin file path validation in extraction command

p1utoze committed:
refactor: add warning message for cryptographic wipe command and fix typo

p1utoze committed:
update: replace if checks with try/except blocks in auto wipe method

p1utoze committed:
update: add method for interactive veracrypt volume creation through CLI and refactor code

p1utoze commented:
> It is not a requirement to test only on SSDs. You can proceed with testing on USB.

I've tested on my USB stick with extraction and sanitization commands. The extraction command works well both as interactive mode as well as `--options` flag mode. 

The sanitization command fails with ATA, SATA and NVMe secure erase methods falling back to crypto-wipe (on auto-wipe selection). Automatic encryption option through Veracrypt worked creating a volume and mounting on the target device. However, the manual mode didn't work because of subprocess endless wait for input. So, I defined an extra method for a interactive veracrypt mode (same as the official) and tested on it as well.

SuchitReddi commented:
Yes, the sata, ata, and nvme methods work only for supported devices. Cryptographic wipe should work on every device. So that is the expected behaviour.

I will test the script on my device asap and get back to you.

SuchitReddi commented on memorywipe.py:
```diff
@@ -0,0 +1,496 @@
+            click.echo(f"Error: Failed to unmount {self.partition}. Please unmount manually.", fg="red")
+            return False
+        except subprocess.CalledProcessError:
+            click.secho(f"{self.partition} has been succesfully mounted", fg="green")
```
I believe it should be "successfully unmounted" in this line?

p1utoze replied:
Apologies for the typo. It was definitely "successfully unmounted".

SuchitReddi commented on memorywipe.py:
```diff
@@ -0,0 +1,496 @@
+            elif q2 == "a":
+                click.echo("Uses AES, with SHA-512 and makes an NTFS filesystem")
+                click.echo("\nEnter strong password for encrypting. (You don't have to remember it)")
+                strongp = click.prompt("So set a random password with special characters: ", hide_input=True, confirmation_prompt=True)
```
The password set here should be random and strong, and the user doesn't have to remember it. So, set the confirmation_prompt to false here.

SuchitReddi replied:
This is a simple fix, so I will edit it myself.

SuchitReddi commented on memorywipe.py:
```diff
@@ -0,0 +1,496 @@
+        }
+        filesystem = filesystem_dict[click.prompt("Select", default=2, type=click.IntRange(1, 6), show_choices=False)]
+        click.echo("\nEnter strong password for encrypting. (You don't have to remember it)")
+        strongp = click.prompt("So set a random password with special characters: ", hide_input=True, confirmation_prompt=True)
```
The password set here should be random and strong, and the user doesn't have to remember it. So, set the confirmation_prompt to false here.

SuchitReddi replied:
This is a simple fix, so I will edit it myself.

SuchitReddi commented on memorywipe.py:
```diff
@@ -0,0 +1,496 @@
+            
+            click.secho(f"{self.partition} overwritten with one write each of random data and zeroes", fg="bright_white")
+            self.name = click.prompt("Set device name")
+            subprocess.run("sudo", "mkfs.ntfs", "-L", f"{self.name}", f"{self.partition}")
```
This line is causing a type error after setting device name. Please check once.  
This is the traceback:
```
Set device name: test
Traceback (most recent call last):
  File "/home/sherl0ck/memorywipe.py", line 496, in <module>
    main()
  File "/usr/lib/python3/dist-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sherl0ck/memorywipe.py", line 435, in sanitize
    wipe.auto_wipe()
  File "/home/sherl0ck/memorywipe.py", line 400, in auto_wipe
    self.crypt_wipe()
  File "/home/sherl0ck/memorywipe.py", line 62, in crypt_wipe
    self._wipe_disk()
  File "/home/sherl0ck/memorywipe.py", line 182, in _wipe_disk
    subprocess.run("sudo", "mkfs.ntfs", "-L", f"{self.name}", f"{self.partition}")
  File "/usr/lib/python3.11/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/subprocess.py", line 835, in __init__
    raise TypeError("bufsize must be an integer")
TypeError: bufsize must be an integer
```

SuchitReddi committed:
Update memorywipe.py

line 208 fixed typo mounted -> unmounted
line 141, 164 changed confirmation_prompt from True -> False

p1utoze committed:
fix: TypeError in wipe disk method

p1utoze committed:
refactor: memorywipe.py

p1utoze committed:
refactor: memorywipe.py

SuchitReddi commented on memorywipe.py:
```diff
@@ -138,36 +162,41 @@ def _veracrypt_interactive(self):
         if out.returncode:
             click.secho(out.stderr.decode(), fg="yellow")
-        show_out = click.prompt("Do you wish to see the output of the command run?", type=click.Choice(["y", "n"]), default="n")
+        show_out = click.prompt("Do you wish to see the output of the command run?", type=click.Choice(["y", "n"]),
```
Is it possible to remove the choice here, and simply show the output? I think it will be better to show the output.

p1utoze replied:
Isn't the veracrypt volume creation output lengthy?  It may exceed beyond the terminal screen and hence i added it as a choice. If that's not an issue, then i'll remove it right away

SuchitReddi replied:
Let's do one thing, send me a screenshot of the output. We can decide if it is required to hide it or not.

p1utoze replied:
Sure. The output is a small "volume successfully created " message. I'll push the changes with no output prompt.

I actually faced an issue while sanitising. It may be a bug perhaps that during wiping and mounting device, my wifi driver service process was terminated. It turned on back after the process got completed. I am not sure this might occur in other devices as well.

SuchitReddi replied:
Is this happening every time?

p1utoze replied:
> Is this happening every time?

No. Out of 7-8 time sanitization tests I performed on the USB stick, It occured only once.

SuchitReddi replied:
Then it might be something else. But if it repeats again, let me know

p1utoze committed:
update: add c error and show output code irrespective of encryption method

p1utoze committed:
update: add missing code for filesystem creation with manual wipe mode

p1utoze committed:
update: enabled output display for veracrypt wipe command execution progress

p1utoze committed:
refactor: add meaningful messages before execution of major commands

SuchitReddi commented:
Cryptographic wipe is perfect. I tested ata_hdparm(self) (line 287). There are a few issues I observed.

1) "click.Abort" is not really aborting the function (line 226, 324, 339). But "raise click.Abort" does. 
I replaced the click with raise click, and it worked. 

2) In lines 306, 320, 409, grep command is given input as 
' "erase unit" '. But this is not working. I removed the outer single quotes and it is working.

3) chk_freeze is not performing the operation correctly. The original script has this important line:

`sudo hdparm -I $partition | grep -i frozen | grep -i not`

This line will check for "frozen". If it is there, it searches for "not". So what it essentially does is to check for not frozen. Why it was done like this is because the amount of white spaces between not and frozen is variable.

But the python function for this is using:

`if "frozen" not in hdparm`

I don't think the logic is the same for both here.

p1utoze commented:
> Cryptographic wipe is perfect. I tested ata_hdparm(self) (line 287). There are a few issues I observed.
>
> 1. "click.Abort" is not really aborting the function (line 226, 324, 339). But "raise click.Abort" does.
>    I replaced the click with raise click, and it worked.

Oh my bad! I had used `raise Abort` in most of the methods and overlooked the ones you mentioned. Fixing it right away.

> 2. In lines 306, 320, 409, grep command is given input as
>    ' "erase unit" '. But this is not working. I removed the outer single quotes and it is working.
I enclosed it in single quotes for the grep pattern believing that it would parse white space. I'll update this.
> 3. chk_freeze is not performing the operation correctly. The original script has this important line:
> 
> `sudo hdparm -I $partition | grep -i frozen | grep -i not`
> 
> This line will check for "frozen". If it is there, it searches for "not". So what it essentially does is to check for not frozen. Why it was done like this is because the amount of white spaces between not and frozen is variable.

If that's the case,  piping the grep will be a better approach. I'll update the command process to check outputs of both grep patterns. In that way, if the frozen keyword itself doesn't exist, it will raise error or else pipe to the "not" statement.
> 
> But the python function for this is using:
> 
> `if "frozen" not in hdparm`
> 
> I don't think the logic is the same for both here.

p1utoze committed:
refactor: modified command args and exception handling statements

p1utoze committed:
Merge branch 'SuchitReddi:main' into py-codebase

SuchitReddi committed:
fixed assert error, and grep for not frozen

SuchitReddi commented on memorywipe.py:
```diff
@@ -288,7 +288,7 @@ def ata_hdparm(self):
         click.secho("You selected ATA Secure Erase...", fg="magenta")
         click.echo("Starting ATA Secure Erase using hdparm...")
         try:
-            assert _install_tool("hdparm")
```
If assert is present here, it is always leading to the except below and terminating. So, I removed assert. Make necessary changes to the remaining code accordingly.

p1utoze replied:
Yes you're correct! There's no boolean return statement for the assert to work.

SuchitReddi commented on memorywipe.py:
```diff
@@ -325,13 +326,13 @@ def _chk_compat_hdparm(self):
         
     def _chk_freeze(self):
         hdparm_output = subprocess.check_output(["sudo", "hdparm", "-I", f"{self.partition}"], text=True)
-        if "frozen" not in hdparm_output:
```
I tried the piping of multiple grep statements you mentioned in slack, but it did not work. So, I have tried this alternative of searching for not frozen. Even though there are multiple white spaces between not and frozen, this approach is working fine.

SuchitReddi commented on memorywipe.py:
```diff
@@ -362,15 +363,15 @@ def _select_enhance(self):
                     "This mode writes predetermined data patterns set by the manufacturer to all areas including bad blocks")
                 click.echo("Please wait... this may take a long time. At-least:")
                 hdparm = subprocess.run(["sudo", "hdparm", "-I", f"{self.partition}"], capture_output=True)
-                subprocess.check_output(["grep", "-i", "erase unit"], input=hdparm.stdout)
```
I don't know the reason, but here and also in line 373, if check_output is present, the command is not working. Replacing it with run is working.

p1utoze replied:
I have fixed this and other similar commands. Interchanged the run() and check_output() function calls in these lines.

SuchitReddi commented:
After checking for freeze status, if device is frozen, we issue a suspend command. This will result in the system being suspended. But if the user clicks any button before a few minutes, the device status might still be frozen. So as a temporary fix, I added _chk_freeze function twice.
But a more permanent fix would be to continue checking for _chk_freeze using while loop, and if the chk_freeze is saying the device is frozen, we will prompt the user to suspend again by asking y or n. If the device is no longer frozen, we will come out of the loop and continue to the next step.
Is it possible?

p1utoze commented:
> After checking for freeze status, if device is frozen, we issue a suspend command. This will result in the system being suspended. But if the user clicks any button before a few minutes, the device status might still be frozen. So as a temporary fix, I added _chk_freeze function twice. But a more permanent fix would be to continue checking for _chk_freeze using while loop, and if the chk_freeze is saying the device is frozen, we will prompt the user to suspend again by asking y or n. If the device is no longer frozen, we will come out of the loop and continue to the next step. Is it possible?

Instead of while loop, will a recursive call in the function itself be helpful? something like:
```
def _chk_freeze(self):
    hdparm_output = subprocess.check_output(["sudo", "hdparm", "-I", f"{self.partition}"], text=True)
    if "not	frozen" in hdparm_output:
        click.echo("Device is not frozen! Continuing to next step...")
    else:
        click.secho("Device is frozen!", fg="yellow")
        click.echo("A system suspend will suspend the device and start it after a minute unfrozen")
        click.secho("YOUR DEVICE WILL TURN OFF FOR A FEW MINUTES IF YOU SELECT YES. DO NOT PANIC\n")
        suspend = click.prompt("Do you want to suspend your system?", type=click.Choice(["y", "n"]), default='n')
        if suspend == "y":
            subprocess.run(["sudo", "systemctl", "suspend"])
            self._chk_freeze()
        else:
            click.secho("You can't continue with this process without suspending...\nStart again.", fg="red")
            raise click.Abort
```

SuchitReddi commented:
Yes, if you think it will work, we can use this.

p1utoze commented:
> Yes, I will test if this will work or not. If it does, we can use this

Another thing I wanted to mention is should we change the default value of this user prompt to yes? 
```
suspend = click.prompt("Do you want to suspend your system?", type=click.Choice(["y", "n"]), default='n')
```
Because the suspend operation is either ways a mandatory for frozen devices.

p1utoze commented:
> Yes, I will test if this will work or not. If it does, we can use this

Sure! I'll be pushing some changes now.

p1utoze committed:
update: major changes in sanitize and extract commands

SuchitReddi commented:
The better default option for suspend would be no. Suspend is a major change, it's best not to set something like that as the default option.

SuchitReddi committed:
Changed a few default options

p1utoze commented:
> The better default option for suspend would be no. Suspend is a major change, it's best not to set something like that as the default option.

Fair Enough. Is there any more changes required to be made @SuchitReddi ?

SuchitReddi committed:
Update --help
